crates/kitsune_p2p/types/src/tx2/tx2_api.rs

```diff
@@ -467,24 +467,32 @@ impl<C: Codec + 'static + Send + Unpin> Tx2EpHnd<C> {
 
     /// Get an existing connection.
     /// If one is not available, establish a new connection.
-    pub fn get_connection<U: Into<TxUrl>>(
+    pub fn get_connection<U: TryInto<TxUrl, Error = KitsuneError>>(
         &self,
         remote: U,
         timeout: KitsuneTimeout,
     ) -> impl std::future::Future<Output = KitsuneResult<Tx2ConHnd<C>>> + 'static + Send {
-        let remote = remote.into();
+        let ep_hnd = self.0.clone();
         let rmap = self.1.clone();
         let metrics = self.2.clone();
         let local_cert = self.3.clone();
-        let fut = self.0.get_connection(remote.clone(), timeout);
-        async move {
-            let con = fut.await?;
-            Ok(Tx2ConHnd::new(local_cert, con, remote, rmap, metrics))
+
+        match remote.try_into() {
+            Ok(remote) => async move {
+                let con = ep_hnd.get_connection(remote.clone(), timeout).await?;
+                Ok(Tx2ConHnd::new(local_cert, con, remote, rmap, metrics))
+            }
+            .boxed(),
+            err => async move {
+                #[allow(unreachable_code)]
+                err.map(|_| unreachable!() as Tx2ConHnd<C>)
+            }
+            .boxed(),
```

Comment:
huh?

Reply:
Ha, yeah this is weird. It's because the Result::Ok type needs to be mapped from `()` to whatever this fn returns (`Tx2ConHnd<C>`), and this will never be Ok, so that's why it's unreachable. Just about converting the type. Not sure if there's a better way.

Reply:
can't you just actually return the error in the error case?

```rust
Ok(async move {
    match remote.try_into() {
        Ok(remote) => {
            let con = ep_hnd.get_connection(remote.clone(), timeout).await?;
            Ok(Tx2ConHnd::new(local_cert, con, remote, rmap, metrics))
        }
        Err(err) => Err(KitsuneError::other(err)),
    }
    .boxed()
    .into()
})
```

Reply:
or, if you ran into lifetime issues with the `remote` type, you might have to do the try_into a line above:

```rust
let remote = remote.try_into();
async move {
    match remote {
        ...
```

Reply:
Good point, yes it was easy to clear up, I was just in a functional state of mind